### DISCUSSION TITLE
Sqlalchemy session gets hung after few hours of inactivity

### DISCUSSION THREAD

**Discussion by itisguptak** [sqlalchemy/sqlalchemy#8038] [Usage Questions]
Hi Sqlalchemy gurus/peers. I need some help with sqlalchemy ORM implementation. I have written a python asyncio app with postgres backend. I'm using asyncpg dialect as DBAPI. Now the issue is that there is something wrong with the session or the connections. The app works (fetches records) correctly when started but after a while (5-6 hours) the next time I hit the same URL, it just hungs while getting me the records.
All I have to do is keep sending few requests and it works. So it could be that the first time it gets the request, sqlalchemy is not able to get a connection from pool but when retried it starts responding. It usually is the first 1-2 requests that gets hung.
post that it works fine.

There are no errors that I see in logs, it's just that the api takes forever to complete that request.

**Comment by CaselIT** [comment]
Hi,

It seems that the connections are becoming stale or similar after a while. You may want to look into https://docs.sqlalchemy.org/en/14/core/engines.html#sqlalchemy.create_engine.params.pool_recycle to that connections are closed after a certain timeout.

**Reply by itisguptak** [reply]
Hi @CaselIT : I already have pool_recycle=3600 set for my engine instance.

**Reply by itisguptak** [reply]
Another thing I have noted after setting echo=True in engine definition, I see my results are cached as old as 2hours. It means there is no connection getting created because the values on DB doesn't change.

**Reply by CaselIT** [reply]
> Another thing I have noted after setting echo=True in engine definition, I see my results are cached as old as 2hours. It means there is no connection getting created because the values on DB doesn't change.

This is not something that's done by sqlalchemy, so you would have to check whatever caching package you are using for it.
If you refer to a `[cached ...]` in the sqlalchemy logging, that's the sql string, not the result. The query is always re-executed.

>Hi @CaselIT : I already have pool_recycle=3600 set for my engine instance.

not sure if trying to reduce that parameter could help. You may also find useful to enable [echo_pool[(https://docs.sqlalchemy.org/en/14/core/engines.html?highlight=echo%20pool#sqlalchemy.create_engine.params.echo_pool) since the issue seems to be on connection.

**Reply by itisguptak** [reply]
Hi, I had turned on the echo_pool as you had suggested and this what I see in logs.


```
/app/lib/python3.8/site-packages/sqlalchemy/log.py: 117 Connection <AdaptedConnection <asyncpg.connection.Connection object at 0x7f221cb2fe40>> exceeded timeout; recycling

/app/lib/python3.8/site-packages/sqlalchemy/log.py: 117 Exception closing connection <AdaptedConnection <asyncpg.connection.Connection object at 0x7f221cb2fe40>>
Traceback (most recent call last):
File "/app/lib/python3.8/site-packages/sqlalchemy/pool/base.py", line 247, in _close_connection 
    self._dialect.do_close (connection)
File "/app/lib/python3.8/site-packages/sqlalchemy/engine/default.py",line 688, in do_close
    dbapi_connection.close()
File "/app/lib/python3.8/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 730, in close
    self.await_(self._connection.close())
File "/app/lib/python3.8/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 76, in await_only
    return current.driver.switch (awaitable)
File "/app/lib/python3.8/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 129, in greenlet_spawn
    value= await result
File "/app/lib/python3.8/site-packages/asyncpg/connection.py", line 1334, in close
    await self. protocol.close (timeout)
File "asyncpg/protocol/protocol.pyx", line 604, in close
File "/usr/lib64/python3.8/asyncio/selector_events.py", line 848, in _read_ready_data_received
    data = self._sock.recv(self.max_size)
TimeoutError: [Errno 110] Connection timed out
```

Basically, it appears it can detect that connection is timed-out, but for recycling it takes forever (I think 5-10 mins). Any resolution for speeding up the recycle?

In above error, it took almost 5-10 mins for that timed out exception to get logged, so it could be that it took significant time for recycle to complete? Because the app starts working after that.

**Reply by itisguptak** [reply]
To be precise, it took 16 mins to log that timeout exception.

**Reply by CaselIT** [reply]
do you provide some kind of arguments to asyncpg? without some kind of example this seems something related to your deployment.

Can you reproduce also in other environment, like a development machine etc?

**Reply by p3l** [reply]
Did anyone, by chance, resolve this problem? If so, would you mind sharing how?
